### PR TITLE
[GPU] Enable blocked formats in Gather primitive

### PR DESCRIPTION
**Pull request by tuxedcat**
### Details:
 - Supports all blocked formats of (f16/f32/i32/i8/u8) data type.

### Tickets:
 - 57708

### CODE REVIEW

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 1]
```diff
@@ -6,53 +6,837 @@
...
+                    for(int l=0;l<x2;l++)
+                        for(int m=0;m<fsv;m++)
+                            ASSERT_EQ((float)ans[i*f2*y2*x2 + (j*fsv+m)*y2*x2 + k*x2 + l],
+                            (float)float16_to_float32(output_ptr[i*f2/fsv*y2*x2*fsv + j*y2*x2*fsv + k*x2*fsv + l*fsv + m]));
+    }
+    // void TearDown() override {}
+};
+TEST_F(gather8fsv4ym1Fixt, a){}
+TEST_F(gather8fsv4ym1Fixt, b){}
+TEST_F(gather8fsv4ym1Fixt, c){}
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:431]: What is these a,b,c?

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 2]
```diff
@@ -6,53 +6,837 @@
...
+        for (size_t i = 0; i < ans.size(); ++i)
+            ASSERT_EQ((float)ans[i], (float)float16_to_float32(output_ptr[i]));
+        
+    }
+    // void TearDown() override {}
+};
+TEST_F(gather8Fixture1, myfixturetest0){}
+TEST_F(gather8Fixture1, myfixturetest1){}
+TEST_F(gather8Fixture1, myfixturetest2){}
+TEST_F(gather8Fixture1, myfixturetest3){}
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:804]: For testing various layouts, I think you can use `TEST_P`, instead of duplicating code.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 3]
```diff
@@ -6,53 +6,837 @@
...
+        //     datbfzyx[i*f0*z0*y0*x0 + (j*fsv+n)*z0*y0*x0 + k*y0*x0 + l*x0 + m]
+        //     = dat[i*f0/fsv*z0*y0*x0*fsv + j*z0*y0*x0*fsv + k*y0*x0*fsv + l*x0*fsv + m*fsv + n];
+        // for(int i=0;i<b1;i++)
+        // for(int j=0;j<f1/fsv;j++)
+        // for(int k=0;k<z1;k++)
+        // for(int l=0;l<y1;l++)
+        // for(int m=0;m<x1;m++)
+        // for(int n=0;n<fsv;n++)
+        //     indbfzyx[i*f1*z1*y1*x1 + (j*fsv+n)*z1*y1*x1 + k*y1*x1 + l*x1 + m]
+        //     = ind[i*f1/fsv*z1*y1*x1*fsv + j*z1*y1*x1*fsv + k*y1*x1*fsv + l*x1*fsv + m*fsv + n];
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:156]: Please remove comment

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 4]
```diff
@@ -6,53 +6,837 @@
...
+        topology.add(gather("gather", "InputDictionary", "InputText", axis, {b2,f2,y2,x2}, batch_dim, true, "", cldnn::padding(), fmt));
+
+        network network(engine, topology);
+        network.set_input_data("InputDictionary", input0);
+        network.set_input_data("InputText", input1);
+
+        auto output = network.execute().at("gather").get_memory();
+        cldnn::mem_lock<uint16_t> output_ptr(output, get_test_stream());
+        
+        auto datbfyx=dat;
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:626]: Please use readable name..

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/include/intel_gpu/primitives/gather.hpp [hunk 5]
```diff
@@ -37,12 +37,14 @@ struct gather : public primitive_base<gather> {
            const int64_t batch_dim = 0,
            const bool support_neg_ind = false,
            const primitive_id& ext_prim_id = "",
-           const padding& output_padding = padding())
+           const padding& output_padding = padding(),
+           const format fmt = format::any)
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/gather.hpp]: Fixed.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 6]
```diff
@@ -6,53 +6,837 @@
...
 #include <intel_gpu/primitives/input_layout.hpp>
 #include <intel_gpu/primitives/gather.hpp>
+#include "ngraph/runtime/reference/gather.hpp"
 
 #include <cstddef>
 
 using namespace cldnn;
 using namespace ::tests;
 
+class gather8LargeFlatFixt : public ::testing::Test {
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: I meant fixture for flat memory format test.  Updated to parameterized version & fixed.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 1 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Updated to parameterized version & fixed.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 2 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Updated to parameterized version.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 3 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Fixed.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 4 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Sorry... fixed.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 7]
```diff
@@ -6,53 +6,837 @@
...
+protected:
+    static const format::type fmt = format::bfzyx;
+    static const int fsv=1;
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Fixed.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 8]
```diff
@@ -46,11 +46,24 @@ ParamsKey GatherKernelRef::GetSupportedKey() const {
...
+    k.EnableInputLayout(DataLayout::b_fs_yx_fsv16);
+    k.EnableInputLayout(DataLayout::b_fs_yx_fsv32);
+    k.EnableInputLayout(DataLayout::b_fs_zyx_fsv16);
+    k.EnableInputLayout(DataLayout::b_fs_zyx_fsv32);
+    k.EnableInputLayout(DataLayout::bs_fs_yx_bsv16_fsv16);
+    k.EnableOutputLayout(DataLayout::bfyx);
+    k.EnableOutputLayout(DataLayout::bfzyx);
     k.EnableOutputLayout(DataLayout::bfwzyx);
+    k.EnableOutputLayout(DataLayout::b_fs_yx_fsv4);
+    k.EnableOutputLayout(DataLayout::b_fs_yx_fsv16);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:61]: Is there any layout that is not supported? Can it be EnableAllLayouts?

File: src/plugins/intel_gpu/tests/CMakeLists.txt [hunk 9]
```diff
@@ -47,7 +47,9 @@ target_link_libraries(${TARGET_NAME} PRIVATE openvino_intel_gpu_graph
 target_include_directories(${TARGET_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}
                                                   ${CMAKE_CURRENT_SOURCE_DIR}/test_utils/
                                                   $<TARGET_PROPERTY:openvino_intel_gpu_kernels,INTERFACE_INCLUDE_DIRECTORIES>
-                                                  $<TARGET_PROPERTY:openvino_intel_gpu_runtime,INTERFACE_INCLUDE_DIRECTORIES>)
+                                                  $<TARGET_PROPERTY:openvino_intel_gpu_runtime,INTERFACE_INCLUDE_DIRECTORIES>
+                                                  $<TARGET_PROPERTY:openvino_intel_gpu_runtime,INTERFACE_INCLUDE_DIRECTORIES>
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/CMakeLists.txt:51]: added twice

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 10]
```diff
@@ -6,53 +6,249 @@
...
 
 #include <cstddef>
 
 using namespace cldnn;
 using namespace ::tests;
 
+#define GATHER8_TEST_4D_FORMATS \
+    format::type::bfyx, format::type::b_fs_yx_fsv4, format::type::b_fs_yx_fsv16, format::type::b_fs_yx_fsv32
+#define GATHER8_TEST_5D_FORMATS format::type::bfzyx, format::type::b_fs_zyx_fsv16, format::type::b_fs_zyx_fsv32
+
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:19]: Could you add double blocked formats too?

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 11]
```diff
@@ -89,55 +284,43 @@ TEST(gather8_gpu_fp16, d323_axisY_bdim_m1) {
...
-
-        103.f, 104.f,  99.f,  100.f, 97.f,  98.f,
-        111.f, 112.f, 107.f, 108.f, 105.f, 106.f,
-        119.f, 120.f, 115.f, 116.f, 113.f, 114.f,
-
-        125.f, 126.f, 121.f, 122.f, 123.f, 124.f,
-        133.f, 134.f, 129.f, 130.f, 131.f, 132.f,
-        141.f, 142.f, 137.f, 138.f, 139.f, 140.f
+    std::vector<FLOAT16> expected_results = {
+       1,2,9,10,17,18,31,32,35,36,41,42,51,52,59,60,67,68,77,78,81,82,95,96,103,104,107,108,113,114,125,126,129,130,139,140
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:309]: Is this passed lint check?

Comment (yeonbok) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:309]: Is this passed lint check?

Comment (yeonbok) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:309]: BTW is this needed?

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:24]: Enabled all format.

Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:24]: Enabled all format.

File: src/plugins/intel_gpu/tests/CMakeLists.txt (hunk 9 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/CMakeLists.txt]: Fixed.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:67]: Enabled all format.

[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:67]: I meant that you can use EnableAllInputLayout or EnableAllOutputLayout instead of adding one by one, if you can ensure that all layouts are supported by this kernel actually.

[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Fixed

[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: This PR guarantees only blocked formats(Includes fs_b_yx_fsv32) works. Planar formats which is not the bfyx order like yxfb won't work correctly in this PR(Note: originally not supported). I'll revert enabling all to enabling each format.

[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:71]: Reverted

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather.cpp [hunk 12]
```diff
@@ -115,6 +115,84 @@ attach_gather_impl::attach_gather_impl() {
...
+        std::make_tuple(data_types::i8, format::bs_fs_yx_bsv16_fsv16),
+        std::make_tuple(data_types::u8, format::bs_fs_yx_bsv16_fsv16),
+
+        std::make_tuple(data_types::f32, format::bs_fs_yx_bsv32_fsv16),
+        std::make_tuple(data_types::f16, format::bs_fs_yx_bsv32_fsv16),
+        std::make_tuple(data_types::i32, format::bs_fs_yx_bsv32_fsv16),
+        std::make_tuple(data_types::i8, format::bs_fs_yx_bsv32_fsv16),
+        std::make_tuple(data_types::u8, format::bs_fs_yx_bsv32_fsv16),
+
+        std::make_tuple(data_types::f32, format::bs_fs_yx_bsv32_fsv32),
```
Comment (jade-cho) [inline src/plugins/intel_gpu/src/graph/impls/ocl/gather.cpp:185]: fsv32 format must couple with 8bit interger data type.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 13]
```diff
@@ -102,16 +145,12 @@ static inline std::vector<std::string> GetOrder(size_t size) {
...
         idx_order = {"b", "f", "w", "z", "y", "x"};
     }
-
     return idx_order;
 }
 
 static std::string GetDictionaryIndexOrder(const gather_params& params, size_t axis) {
     std::vector<std::string> idx_order = GetOrder(params.outputs[0].GetDims().size());
 
-    const std::string input_axis_index_macro = "INPUT_AXIS_INDEX";
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:112]: Reverted relevant changes.

**Review by jade-cho — APPROVED** [review]

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 14]
```diff
@@ -102,15 +147,14 @@ static inline std::vector<std::string> GetOrder(size_t size) {
...
         idx_order = {"b", "f", "w", "z", "y", "x"};
     }
-
     return idx_order;
 }
 
 static std::string GetDictionaryIndexOrder(const gather_params& params, size_t axis) {
-    std::vector<std::string> idx_order = GetOrder(params.outputs[0].GetDims().size());
+    const std::string& input_axis_index_macro = "INPUT_AXIS_INDEX";
+    const std::string& zero_val = "0";
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:155]: Why did you flip the sequence of these three lines?

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 15]
```diff
@@ -159,23 +203,10 @@ CommonDispatchData GatherKernelRef::SetDefault(const gather_params& params, cons
...
-        dispatchData.gws = {output.X().v * output.Y().v, output.Z().v * output.W().v, output.Feature().v * output.Batch().v};
-        dims_by_gws = {{Tensor::DataChannelName::X, Tensor::DataChannelName::Y},
-                       {Tensor::DataChannelName::Z, Tensor::DataChannelName::W},
-                       {Tensor::DataChannelName::FEATURE, Tensor::DataChannelName::BATCH}};
-    }
-
+    dispatchData.gws = {output.X().v * output.Y().v, output.Z().v * output.W().v, output.Feature().v * output.Batch().v};
+    dims_by_gws = {{Tensor::DataChannelName::X, Tensor::DataChannelName::Y},
+                   {Tensor::DataChannelName::Z, Tensor::DataChannelName::W},
+                   {Tensor::DataChannelName::FEATURE, Tensor::DataChannelName::BATCH}};
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:209]: Why did you change this logic? This will change performance in 4d data where Z=W=0.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 16]
```diff
@@ -8,10 +8,334 @@
...
 
 using namespace cldnn;
 using namespace ::tests;
 
+std::ostream& operator<<(std::ostream& os, FLOAT16 x) {
+    return os << float(x);
+}
+
+template <class T>
+int dim(const T& a) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:21]: what about changing name to `get_effective_dim` or `get_not_one_dim`?

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 17]
```diff
@@ -8,10 +8,334 @@
...
+    int ret = a.size();
+    while (ret - 1 >= 0 && a[ret - 1] == 1)
+        ret--;
+    return ret;
+};
+
+template <class T>
+int mult(const T& a) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:29]: I think `multiplication` is a misleading name. Please use proper one. We have similar function for `struct layout`: `get_linear_size`

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 18]
```diff
@@ -8,10 +8,334 @@
...
+                                         format::type,         // format of input0
+                                         format::type,         // format of input1
+                                         std::array<int, 5>,   // bfzyx0
+                                         std::array<int, 5>>;  // bfzyx1
+
+// index 0 is input0
+// index 1 is input1
+// index 2 is output
+template <class T_dat, class T_ind, data_types T_dat_dt, data_types T_ind_dt>
+class gather8_test_4d : public ::testing::TestWithParam<gather8_test_4d_param> {
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:52]: I think we can merge two functions `gather8_test_4d` and `gather8_test_5d` by using `vector` instead of `array`?

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 19]
```diff
@@ -8,10 +8,334 @@
...
+using gather8_test_4d_f16i32 = gather8_test_4d<FLOAT16, int, data_types::f16, data_types::i32>;
+TEST_P(gather8_test_4d_f16i32, gather8_test_4d_f16i32) {}
+INSTANTIATE_TEST_SUITE_P(gather8_4d_bd0_d4_i1,
+                         gather8_test_4d_f16i32,
+                         testing::Combine(testing::Values(0),
+                                          testing::Values(0),  //[batch_dim,dim(dict))
+                                          testing::Values(format::type::bfyx),
+                                          testing::Values(format::type::bfyx),
+                                          testing::Values(std::array<int, 4>{5, 44, 7, 8}),
+                                          testing::Values(std::array<int, 4>{4, 1, 1, 1})));
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:232]: I think you can merge many `INSTANTIATE` into smaller list with `testing::Values`

File: src/plugins/intel_gpu/src/graph/gather.cpp [hunk 20]
```diff
@@ -21,17 +21,44 @@ layout gather_inst::calc_output_layout(gather_node const& node) {
...
-    if (node.has_fused_primitives()) {
-        output_type = node.get_fused_output_layout().data_type;
+    format output_format = input_layout.format;
+    if (dims_converted.size() == 5) {
+        switch (input_layout.format) {
+        case format::bfyx:
+            output_format = format::get_default_format(dims_converted.size());
+            break;
+        case format::b_fs_yx_fsv4:
+            // output_format = format::b_fs_zyx_fsv4;
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/graph/gather.cpp]: Done.

File: src/plugins/intel_gpu/src/graph/gather.cpp [hunk 21]
```diff
@@ -21,17 +21,41 @@ layout gather_inst::calc_output_layout(gather_node const& node) {
...
+        default:
+            break;
+        }
     }
-
-    return layout{output_type, output_format, tensor(output_format, dims_converted)};
+    auto output_type = input_layout.data_type;
+    return layout{output_type,
+                  output_format,
+                  tensor(format::get_default_format(dims_converted.size()), dims_converted)};
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/graph/gather.cpp:58]: Discussion and conclusion about calc_out_layout(tensor redundant axis issue) of #11652 was that. Changes of line 27~53 are to support dimension promotion(cross dimension?). Changes of line 55~58 are to support fs_b_yx_fsv32 format.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 14 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:155]: It didn't have any special meaning, just changed while modifying some codes. I will revert that changes soon.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 15 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:209]: At this time, I did not know how to obtain the dimension of the block format. I saw that other kernels using such gws and I thought this method was concise. I understand that modular operation is wasted when in 4D. It seems it can be handled with get_default_format(format) function. I will fix it soon.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 19 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp:232]: Passing multiple values to testing::Values() generates testcases as Cartesian product, which conflicts with feedback that there are too many test cases.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp [hunk 22]
```diff
@@ -8,10 +8,334 @@
...
+INSTANTIATE_TEST_SUITE_P(gather8_4d_bd0_d2_i2,
+                         gather8_test_4d_f16i32,
+                         testing::Combine(testing::Values(0),
+                                          testing::Values(1),  //[batch_dim,dim(dict))
+                                          testing::Values(format::type::b_fs_yx_fsv4),
+                                          testing::Values(format::type::b_fs_yx_fsv16),
+                                          testing::Values(std::array<int, 4>{8, 67, 1, 1}),
+                                          testing::Values(std::array<int, 4>{4, 56, 1, 1})));
+INSTANTIATE_TEST_SUITE_P(gather8fs_b_yx_fsv32,
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Done

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Done.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 23]
```diff
@@ -132,9 +176,9 @@ static std::string GetDictionaryIndexOrder(const gather_params& params, size_t a
 }
 
 static std::string GetIndecesIdxOrder(const gather_params& params, size_t axis, int64_t batch_dim) {
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Done.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 14 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Done.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 15 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Done.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 16 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Done.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 17 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Done.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 18 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Done.

File: src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp (hunk 19 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/gather_gpu_test.cpp]: Added some disabled testcases for future usage.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:74]: Test fails(segmentation fault) in byxf. Previous kernel is written for only default planar formats and I worked only blocked formats, so non-default planar format currently not works.

[the remark this answers is not in the document]
Comment (isanghao) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:74]: Could you check why the test is failing?

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 24]
```diff
@@ -159,24 +204,32 @@ CommonDispatchData GatherKernelRef::SetDefault(const gather_params& params, cons
     auto out_layout = params.outputs[0].GetLayout();
     std::vector<std::vector<Tensor::DataChannelName>> dims_by_gws;
 
-    if (out_layout == DataLayout::bfyx) {
+
+    // int rank = cldnn::format(out_layout.).dimension();
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:208]: Please remove unnecessary code

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Updated to support non-planar format [447c417](https://github.com/openvinotoolkit/openvino/pull/11486/commits/447c4172f319ccd211f405c869b66e4b83a039f7)

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 24 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Done.

**Review by isanghao — APPROVED** [review]
